rust/src/bin/main.rs

```diff
@@ -40,17 +48,47 @@ async fn main() {
     // is correctly initialized
     sodiumoxide::init().unwrap();
 
-    let (state_machine, requests_tx, event_subscriber) =
-        StateMachine::<Traced<Request>>::new(pet_settings, mask_settings, model_settings).unwrap();
+    #[cfg(feature = "metrics")]
+    let (metrics_sender, metrics_handle) = {
+        let (metrics_service, metrics_sender) =
+            MetricsService::new("http://influxdb:8086", "metrics");
+        (
+            metrics_sender,
+            tokio::spawn(async { run_metric_service(metrics_service).await }),
+        )
+    };
+
+    let (state_machine, requests_tx, event_subscriber) = StateMachine::<Traced<Request>>::new(
+        pet_settings,
+        mask_settings,
+        model_settings,
+        #[cfg(feature = "metrics")]
+        metrics_sender,
+    )
+    .unwrap();
     let fetcher = services::fetcher(&event_subscriber);
     let message_handler = services::message_handler(&event_subscriber, requests_tx);
 
+    let state_machine_handle = tokio::spawn(async { state_machine.run().await });
+
     tokio::select! {
-        _ = state_machine.run() => {
-            println!("shutting down: Service terminated");
-        }
         _ = rest::serve(api_settings.bind_address, fetcher, message_handler) => {
-            println!("shutting down: REST server terminated");
+            warn!("shutting down: REST server terminated");
         }
+        _ =  signal::ctrl_c() => {}
+    }
+
+    // The moment the message_handler is dropped, the sender half of the request channel is also
+    // dropped, which means that the state machine shuts itself down.
+    warn!("shutting down service");
+    let _ = state_machine_handle.await;
```

Comment:
The state machine can only be terminated if all request senders have been dropped. Since only the HTTP service has a sender, there is no way the state machine can terminate earlier than the HTTP service. Of course there is also the possibility of panic but in this case the whole program would just crash.

Reply:
> The state machine can only be terminated if all request senders have been dropped

As of today, this is true, but we may introduce other termination conditions in the future. And in that case, we might run into the case @finiteprods mentioned. But I agree about the clean shutdown issue. Maybe we could have a custom drop implementation that performs the shutdown?

Reply:
I agree. I moved the state machine back into the `select!`. I really like the idea to replace the shutdown state with a custom drop implementation👍